src/stk/image/volume.cpp

```diff
@@ -315,6 +339,35 @@ void Volume::release()
     _strides[2] = 0;
     _origin = { 0, 0, 0 };
     _spacing = { 1, 1, 1 };
+    _direction.diagonal({ 1, 1, 1 });
+    _metadata.reset();
+}
+std::vector<std::string> Volume::get_metadata_keys(void) const
+{
+    std::vector<std::string> keys;
+    keys.reserve(_metadata->size());
+    for (auto const& [k, _] : *_metadata) {
+        keys.push_back(k);
+    }
+    return keys;
+}
+std::string Volume::get_metadata(const std::string& key) const
+{
+    try {
+        return _metadata->at(key);
+    }
+    catch (const std::out_of_range&) {
+        FATAL() << "Metadata '" << &key << "' not found";
+    }
+}
+void Volume::set_metadata(const std::string& key, const std::string& value)
+{
+    if (_metadata.use_count() > 1) {
+        auto new_metadata = std::make_shared<MetaDataDictionary>();
+        new_metadata->insert(_metadata->begin(), _metadata->end());
+        _metadata = new_metadata;
```

Comment:
Did you mean something like this? I am kinda exhausted, so I hope I did not forget to fix all occurrences...

Reply:
Yes, exactly. Could probably add a disclaimer to the member in the header to not access by other means than these methods.

One thing though; this is not thread-safe, but I don't think this will be an issue since volumes are constricted to a single thread typically. But I may be wrong.

Reply:
Well, correct me if I am wrong, but I think none of the getters/setters are currently thread-safe. The additional annoying part of the COW is that to make it thread-safe we need a critical section inside the constructors I think. Sounds like a very interesting way to generate deadlocks if not done carefully.

Guess we can add a mutex to `Volume` if we find it necessary, but ATM maybe it is better to just add a comment in the header and move on.

Reply:
BTW, I think the COW is a neat feature of course, but likely the dictionary will be empty most of the times, and when metadata is present it will be a handful of small values (study and patient IDs, scanner parameters, tracer info et similia), so even making deep copies all the time probably won't hurt in a noticeable way. Unless someone decides to store the full text of *Paradise Lost* as NIFTI file metadata, but that's pretty unlikely I guess...

Reply:
Correct and this is kinda the same thing you would expect from something like `std::vector`. I can't come up with any obvious cases right now, but I'm talking about specifically this COW thing, where we check the `user_count`. 

As a user you wouldn't expect to be required to make sure a volume and its subvolumes are thread-safe for instance (more than the actual image data), but yeah, finding a case where you get a false `use_count() == 1` is hard so this is probably a non-issue.

This could be quite easily be solved with an atomic shared_ptr, but I don't think there's a real need for it.

Reply:
Not 100% sure using an atomic shared_ptr is enough to have full thread-safety, since in `set_metadata` we have a critical section that spans over multiple instructions. But I have not thought through all the details.

Reply:
atomic shared_ptr should solve the problem I'm speaking of though. 

A bad example: You have volume A on two threads, thread 1 starts writing to metadata and at the exact same time thread two makes a subvolume of A. This is kinda invalid on its own since the user is responsible for thread-safety in this case. 

Extended example: Thread 2 does not hold Volume A, but a subvolume of A called B, which makes more sense. Then at the same time A is writing to metadata you create a new subvolume of B.

It just dawn on me that this is a fundemental flaw in `shared_ptr` so they probably already used atomic reference counting.